action.yaml

```diff
@@ -257,62 +313,139 @@ runs:
+        cd ${{ inputs.root-folder }}
+        branch=$(echo "${GITHUB_REF#refs/heads/}")
+        echo "Branch pushed to: $branch"
+        git fetch origin $branch
+        files=$(git diff --name-only ${{ github.event.before }} ${{ github.event.after }})
+        echo "files changed: $files"
+        dags_only=1
+
+        skip_deploy=1
+
+        # This is to ensure that the we are backward compatible for the case when
+        # the git repo only contains astro project and root-folder is not set in the GHA
+        root_folder="${{ inputs.root-folder }}"
+        if [[ $root_folder == "./" ]]; then
+          root_folder=""
         fi
 
-        if [[ ${{steps.deployment-preview.outputs.SKIP_DEPLOY}} == true ]]; then
+        # This for loop checks for following cases:
+        # 1. If no file is part of the input root folder, then it skips deploy
+        # 2. If any file is not part of the dags folder, then it triggers a full image build
+        # 3. If all files are part of the dags folder and input root folder, then it triggers a DAG-only deploy
+        for file in $files; do
+          if [[ $file =~ ^"$root_folder".* ]]; then
+            echo $file is part of the input root folder
+            skip_deploy=0
+            if [[ $file != *"dags/"* ]]; then
+              echo $file is not a DAG, triggering a full image build
+              dags_only=0
+              break
+            fi
+          fi
+        done
+
+        # Note: the order of these following checks is important to ensure that we skip/trigger deploy correctly in following cases:
+        # 1. When there is no change in the input root folder we should skip deploy, but not when it's deployment preview create action
+        # 2. When user has passed a custom image then we would need to deploy the image
+        # 3. When the action is deployment preview delete, then we should skip any form of deploy
+        if [[ $skip_deploy == 1 ]]; then
           # skip all deploy steps
           dags_only=2
         fi
 
-        if [[ ${{ inputs.deploy-image }} == true ]]; then
-          # make sure image and DAGs deploys because deploy-image is true
+        # check if user has passed a custom image or the action has created a new deployment preview, then we would need to deploy the image
+        if [[ ${{ inputs.image-name }} != "no-custom-image" || ${{ steps.dag-deploy-enabled.outputs.DAG_DEPLOY_ENABLED }} == false || ${{ steps.deployment-preview.outputs.IMAGE_DEPLOY_ONLY }} == true ]]; then
           dags_only=0
         fi
 
+        if [[ ${{ steps.deployment-preview.outputs.SKIP_DEPLOY }} == true ]]; then
+          # skip all deploy steps
+          dags_only=2
+        fi
+
         echo "DAGS_ONLY=$dags_only" >> $GITHUB_OUTPUT
+        echo ::endgroup::
       shell: bash
       id: deployment-type
     # If only DAGs changed and dag deploys is enabled, do a DAG-only deploy
-    - name: setup deploy options
+    - name: Setup image or dags deploy options
```

Comment:
We could merge them, but both are decently sized bash scripts, so we should avoid merging them unless we see a value addition by doing so otherwise it would be another very long script to manage. This function is well-scoped, and the other one isn't, and I agree with that.